README.md
Resemble.js
==========

Analyse and compare images with Javascript and HTML5. [Resemble.js Demo](http://huddle.github.com/Resemble.js/)

![Two image diff examples side-by-side, one pink, one yellow.](https://raw.github.com/Huddle/Resemble.js/master/demoassets/readmeimage.jpg "Visual image comparison")

### Get it

`npm install resemblejs`

`bower install resemblejs`

### Example

Retrieve basic analysis on image.

```javascript
var api = resemble(fileData).onComplete(function(data){
	console.log(data);
	/*
	{
	  red: 255,
	  green: 255,
	  blue: 255,
	  brightness: 255
	}
	*/
});
```

Use resemble to compare two images.

```javascript
var diff = resemble(file).compareTo(file2).ignoreColors().onComplete(function(data){
	console.log(data);
	/*
	{
	  misMatchPercentage : 100, // %
	  isSameDimensions: true, // or false
	  dimensionDifference: { width: 0, height: -1 }, // defined if dimensions are not the same
	  getImageDataUrl: function(){}
	}
	*/
});
```

Scale second image to dimensions of the first one:
```javascript
//diff.useOriginalSize();
diff.scaleToSameSize();
```

You can also change the comparison method after the first analysis.

```javascript
// diff.ignoreNothing();
// diff.ignoreColors();
diff.ignoreAntialiasing();
```


And change the output display style.

```javascript
resemble.outputSettings({
  errorColor: {
    red: 255,
    green: 0,
    blue: 255
  },
  errorType: 'movement',
  transparency: 0.3,
  largeImageThreshold: 1200
});
// resembleControl.repaint();
```

By default, the comparison algorithm skips pixels when the image width or height is larger than 1200 pixels. This is there to mitigate performance issues.

You can switch this modify this behaviour by setting the `largeImageThreshold` option to a different value. Set it to **0** to switch it off completely.

--------------------------------------

Created by [James Cryer](http://github.com/jamescryer) and the Huddle development team.


CHANGELOG.md
#0.2.0

  * added support for jpeg

#0.1.0

  * wrong release - kept back

#0.0.4

  * rename package name to node-resemble-js

#0.0.3

  * added ignoreRectangles feature

<a name="0.0.0"></a>
#0.0.0
  * initial commit (started form resemble sha 5ce379b)
  * rewrite resence to node.js (with only pngjs dependency)
  * added exampl


resemble.spec.js
'use strict';

describe('node-resemble.js', function() {
  var EXAMPLE_LARGE_IMAGE = 'example/LargeImage.png';
  var EXAMPLE_SMALL_IMAGE = 'example/SmallImage.png';
  var OPTIMISATION_SKIP_STEP = 6;
  var DEFAULT_LARGE_IMAGE_THRESHOLD = 1200;

  var resemble = require('./resemble.js');

  describe('largeImageThreshold', function() {
    describe('when unset', function() {
      describe('when ignoreAntialiasing is enabled', function() {
        it('skips pixels when a dimension is larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels when both dimensions are smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        it('does not skip pixels when a dimension is larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels when both dimensions are smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    describe('when explicitly set', function() {
      afterAll(function() {
        resemble.outputSettings({largeImageThreshold: DEFAULT_LARGE_IMAGE_THRESHOLD});
      });

      describe('when ignoreAntialiasing is enabled', function() {
        it('skips pixels on images with a dimension larger than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 999});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension equal to the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1000});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with both dimensions smaller than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1001});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        it('does not skip pixels on images with a dimension larger than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 999});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension equal to the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1000});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with both dimensions smaller than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1001});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    describe('when set to a falsy value', function() {
      beforeEach(function() {
        resemble.outputSettings({largeImageThreshold: 0});
      });

      afterAll(function() {
        resemble.outputSettings({largeImageThreshold: DEFAULT_LARGE_IMAGE_THRESHOLD});
      });

      describe('when ignoreAntialiasing is enabled', function() {
        it('does not skip pixels on images with a dimension larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        it('does not skip pixels on images with a dimension larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    function expectPixelsToBeSkipped(image, step) {
      expect(getPixelForLocation(image, 1, step - 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, 1, step).alpha).toBe(0);
      expect(getPixelForLocation(image, 1, step + 1).alpha).not.toBe(0);

      expect(getPixelForLocation(image, step - 1, 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, 1).alpha).toBe(0);
      expect(getPixelForLocation(image, step + 1, 1).alpha).not.toBe(0);

      expect(getPixelForLocation(image, step, step).alpha).toBe(0);
    }

    function expectPixelsNotToBeSkipped(image, step) {
      expect(getPixelForLocation(image, 1, step).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, step).alpha).not.toBe(0);
    }
  });

  function getLargeImageComparison() {
    return resemble(EXAMPLE_LARGE_IMAGE).compareTo(EXAMPLE_LARGE_IMAGE);
  }

  function getSmallImageComparison() {
    return resemble(EXAMPLE_SMALL_IMAGE).compareTo(EXAMPLE_SMALL_IMAGE);
  }

  function getPixelForLocation(image, x, y) {
    var index = (image.width * y + x) << 2;
    return {
      red: image.data[index],
      green: image.data[index + 1],
      blue: image.data[index + 2],
      alpha: image.data[index + 3]
    };
  }
});


demoassets/main.js
$(function(){
	var $target = $('#drop-zone');

	function dropZone($target, onDrop){
		$target.
			bind('dragover', function(){
				$target.addClass( 'drag-over' );
				return false;
			}).
			bind("dragend", function () {
				$target.removeClass( 'drag-over' );
				return false;
			}).
			bind("dragleave", function () {
				$target.removeClass( 'drag-over' );
				return false;
			}).
			bind("drop", function(event) {
				var file = event.originalEvent.dataTransfer.files[0];

				event.stopPropagation();
				event.preventDefault();

				$target.removeClass( 'drag-over' );

				var droppedImage = new Image();
				var fileReader = new FileReader();

				fileReader.onload = function (event) {
					droppedImage.src = event.target.result;
					$target.html(droppedImage);
				};

				fileReader.readAsDataURL(file);

				onDrop(file);
			});
	}

	dropZone($target, function(file){

		resemble(file).onComplete(function(data){
			$('#image-data').show();
			$('#red').css('width',data.red+'%');
			$('#green').css('width',data.green+'%');
			$('#blue').css('width',data.blue+'%');
			$('#alpha').css('width',data.alpha+'%');
			$('#brightness').css('width',data.brightness+'%');
		});

	});

	function onComplete(data){
		var time = Date.now();
		var diffImage = new Image();
		diffImage.src = data.getImageDataUrl();

		$('#image-diff').html(diffImage);

		$(diffImage).click(function(){
			window.open(diffImage.src, '_blank');
		});

		$('.buttons').show();

		if(data.misMatchPercentage == 0){
			$('#thesame').show();
			$('#diff-results').hide();
		} else {
			$('#mismatch').text(data.misMatchPercentage);
			if(!data.isSameDimensions){
				$('#differentdimensions').show();
			} else {
				$('#differentdimensions').hide();
			}
			$('#diff-results').show();
			$('#thesame').hide();
		}
	}

	var file1;
	var file2;
	var resembleControl;

	dropZone($('#dropzone1'), function(file){
		file1 = file;
		if(file2){
			resembleControl = resemble(file).compareTo(file2).onComplete(onComplete);
		}
	});

	dropZone($('#dropzone2'), function(file){
		file2 = file;
		if(file1){
			resembleControl = resemble(file).compareTo(file1).onComplete(onComplete);
		}
	});

	var buttons = $('.buttons button');

	buttons.click(function(){
		var $this = $(this);

		$this.parent('.buttons').find('button').removeClass('active');
		$this.addClass('active');

		if($this.is('#raw')){
			resembleControl.ignoreNothing();
		}
		else
		if($this.is('#less')){
			resembleControl.ignoreLess();
		}
		if($this.is('#colors')){
			resembleControl.ignoreColors();
		}
		else
		if($this.is('#antialising')){
			resembleControl.ignoreAntialiasing();
		}
		else
		if($this.is('#same-size')){
			resembleControl.scaleToSameSize();
		}
		else
		if($this.is('#original-size')){
			resembleControl.useOriginalSize();
		}
		else
		if($this.is('#pink')){
			resemble.outputSettings({
				errorColor: {
					red: 255,
					green: 0,
					blue: 255
				}
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#yellow')){
			resemble.outputSettings({
				errorColor: {
					red: 255,
					green: 255,
					blue: 0
				}
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#flat')){
			resemble.outputSettings({
				errorType: 'flat'
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#movement')){
			resemble.outputSettings({
				errorType: 'movement'
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#flatDifferenceIntensity')){
			resemble.outputSettings({
				errorType: 'flatDifferenceIntensity'
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#movementDifferenceIntensity')){
			resemble.outputSettings({
				errorType: 'movementDifferenceIntensity'
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#opaque')){
			resemble.outputSettings({
				transparency: 1
			});
			resembleControl.repaint();
		}
		else
		if($this.is('#transparent')){
			resemble.outputSettings({
				transparency: 0.3
			});
			resembleControl.repaint();
		}
	});

	(function(){
		var xhr = new XMLHttpRequest();
		var xhr2 = new XMLHttpRequest();
		var done = $.Deferred();
		var dtwo = $.Deferred();

		xhr.open('GET', 'demoassets/People.jpg', true);
		xhr.responseType = 'blob';
		xhr.onload = function(e) {
			done.resolve(this.response);
		};
		xhr.send();

		xhr2.open('GET', 'demoassets/People2.jpg', true);
		xhr2.responseType = 'blob';
		xhr2.onload = function(e) {
			dtwo.resolve(this.response);
		};
		xhr2.send();

		$('#example-images').click(function(){

			$('#dropzone1').html('<img src="demoassets/People.jpg"/>');
			$('#dropzone2').html('<img src="demoassets/People2.jpg"/>');

			$.when(done, dtwo).done(function(file, file1){
				if (typeof FileReader === 'undefined') {
					resembleControl = resemble('demoassets/People.jpg').compareTo('demoassets/People2.jpg').onComplete(onComplete);
				} else {
					resembleControl = resemble(file).compareTo(file1).onComplete(onComplete);
				}
			});

			return false;
		});

	}());

});


example/example.js
var resemble = require('../resemble.js');
var fs = require('fs');

resemble.outputSettings({
  errorColor: {
    red: 155,
    green: 100,
    blue: 155
  },
  errorType: 'movement',
  transparency: 0.6
});

resemble('People.png').compareTo('People2.png')
  //.ignoreAntialiasing()
  //.ignoreColors()
  .onComplete(function(data){
    console.log(data);
    data.getDiffImage().pack().pipe(fs.createWriteStream('diff.png'));
  });

// jpeg comparison
resemble('People.jpg').compareTo('People2.jpg')
  //.ignoreAntialiasing()
  //.ignoreColors()
  .onComplete(function(data){
    console.log(data);
    data.getDiffImage().pack().pipe(fs.createWriteStream('diffjpg.png'));
  });

var fileData1 = fs.readFileSync('People.png');
var fileData2 = fs.readFileSync('People2.png');
resemble(fileData1).compareTo(fileData2)
  //.ignoreAntialiasing()
  //.ignoreColors()
  .ignoreRectangles([[325,170,100,40]])
  .onComplete(function(data){
    console.log('with ignore rectangle:', data);
    data.getDiffImage().pack().pipe(fs.createWriteStream('diffr.png'));
  });


resemble.js
/*
James Cryer / Huddle 2014
URL: https://github.com/Huddle/Resemble.js
*/
'use strict';

var PNG = require('pngjs').PNG;
var fs = require('fs');
var jpeg = require('jpeg-js');

//keeping wrong indentation and '_this' for better diff with origin resemble.js
var _this = {};

	var pixelTransparency = 1;

	var errorPixelColor = { // Color for Error Pixels. Between 0 and 255.
		red: 255,
		green: 0,
		blue: 255,
		alpha: 255
	};

	var errorPixelTransform = {
		flat : function (d1, d2){
			return {
				r: errorPixelColor.red,
				g: errorPixelColor.green,
				b: errorPixelColor.blue,
				a: errorPixelColor.alpha
			}
		},
		movement: function (d1, d2){
			return {
				r: ((d2.r*(errorPixelColor.red/255)) + errorPixelColor.red)/2,
				g: ((d2.g*(errorPixelColor.green/255)) + errorPixelColor.green)/2,
				b: ((d2.b*(errorPixelColor.blue/255)) + errorPixelColor.blue)/2,
				a: d2.a
			}
		}
	};
	
	var errorPixelTransformer = errorPixelTransform.flat;
	var largeImageThreshold = 1200;

	_this['resemble'] = function( fileData ){

		var data = {};
		var images = [];
		var updateCallbackArray = [];

		var tolerance = { // between 0 and 255
			red: 16,
			green: 16,
			blue: 16,
			alpha: 16,
			minBrightness: 16,
			maxBrightness: 240
		};

		var ignoreAntialiasing = false;
		var ignoreColors = false;
    var ignoreRectangles = null;

		function triggerDataUpdate(){
			var len = updateCallbackArray.length;
			var i;
			for(i=0;i<len;i++){
				if (typeof updateCallbackArray[i] === 'function'){
					updateCallbackArray[i](data);
				}
			}
		}

		function loop(x, y, callback){
			var i,j;

			for (i=0;i<x;i++){
				for (j=0;j<y;j++){
					callback(i, j);
				}
			}
		}

		function parseImage(sourceImageData, width, height){

			var pixleCount = 0;
			var redTotal = 0;
			var greenTotal = 0;
			var blueTotal = 0;
			var brightnessTotal = 0;

			loop(height, width, function(verticalPos, horizontalPos){
				var offset = (verticalPos*width + horizontalPos) * 4;
				var red = sourceImageData[offset];
				var green = sourceImageData[offset + 1];
				var blue = sourceImageData[offset + 2];
				var brightness = getBrightness(red,green,blue);

				pixleCount++;

				redTotal += red / 255 * 100;
				greenTotal += green / 255 * 100;
				blueTotal += blue / 255 * 100;
				brightnessTotal += brightness / 255 * 100;
			});

			data.red = Math.floor(redTotal / pixleCount);
			data.green = Math.floor(greenTotal / pixleCount);
			data.blue = Math.floor(blueTotal / pixleCount);
			data.brightness = Math.floor(brightnessTotal / pixleCount);

			triggerDataUpdate();
		}

		function loadImageData( fileData, callback ){

			if (Buffer.isBuffer(fileData)) {
				var png = new PNG();
				png.parse(fileData, function (err, data) {
				  callback(data, data.width, data.height);
				});
			} else {
				var ext = fileData.substring(fileData.lastIndexOf(".")+1);
				if(ext=="png") {
					var png = new PNG();
					fs.createReadStream(fileData)
					  .pipe(png)
					  .on('parsed', function() {
						callback(this, this.width, this.height);
					  });
				}
				if(ext=="jpg" || ext=="jpeg") {
					var jpegData = fs.readFileSync(fileData);
					fileData = jpeg.decode(jpegData, true);
					callback(fileData, fileData.width, fileData.height);
				}
			};
		}

		function isColorSimilar(a, b, color){

			var absDiff = Math.abs(a - b);

			if(typeof a === 'undefined'){
				return false;
			}
			if(typeof b === 'undefined'){
				return false;
			}

			if(a === b){
				return true;
			} else if ( absDiff < tolerance[color] ) {
				return true;
			} else {
				return false;
			}
		}

		function isNumber(n) {
			return !isNaN(parseFloat(n));
		}

		function isPixelBrightnessSimilar(d1, d2){
			var alpha = isColorSimilar(d1.a, d2.a, 'alpha');
			var brightness = isColorSimilar(d1.brightness, d2.brightness, 'minBrightness');
			return brightness && alpha;
		}

		function getBrightness(r,g,b){
			return 0.3*r + 0.59*g + 0.11*b;
		}

		function isRGBSame(d1,d2){
			var red = d1.r === d2.r;
			var green = d1.g === d2.g;
			var blue = d1.b === d2.b;
			return red && green && blue;
		}

		function isRGBSimilar(d1, d2){
			var red = isColorSimilar(d1.r,d2.r,'red');
			var green = isColorSimilar(d1.g,d2.g,'green');
			var blue = isColorSimilar(d1.b,d2.b,'blue');
			var alpha = isColorSimilar(d1.a, d2.a, 'alpha');

			return red && green && blue && alpha;
		}

		function isContrasting(d1, d2){
			return Math.abs(d1.brightness - d2.brightness) > tolerance.maxBrightness;
		}

		function getHue(r,g,b){

			r = r / 255;
			g = g / 255;
			b = b / 255;
			var max = Math.max(r, g, b), min = Math.min(r, g, b);
			var h;
			var d;

			if (max == min){
				h = 0; // achromatic
			} else{
				d = max - min;
				switch(max){
					case r: h = (g - b) / d + (g < b ? 6 : 0); break;
					case g: h = (b - r) / d + 2; break;
					case b: h = (r - g) / d + 4; break;
				}
				h /= 6;
			}

			return h;
		}

		function isAntialiased(sourcePix, data, cacheSet, verticalPos, horizontalPos, width){
			var offset;
			var targetPix;
			var distance = 1;
			var i;
			var j;
			var hasHighContrastSibling = 0;
			var hasSiblingWithDifferentHue = 0;
			var hasEquivilantSibling = 0;

			addHueInfo(sourcePix);

			for (i = distance*-1; i <= distance; i++){
				for (j = distance*-1; j <= distance; j++){

					if(i===0 && j===0){
						// ignore source pixel
					} else {

						offset = ((verticalPos+j)*width + (horizontalPos+i)) * 4;
						targetPix = getPixelInfo(data, offset, cacheSet);

						if(targetPix === null){
							continue;
						}

						addBrightnessInfo(targetPix);
						addHueInfo(targetPix);

						if( isContrasting(sourcePix, targetPix) ){
							hasHighContrastSibling++;
						}

						if( isRGBSame(sourcePix,targetPix) ){
							hasEquivilantSibling++;
						}

						if( Math.abs(targetPix.h - sourcePix.h) > 0.3 ){
							hasSiblingWithDifferentHue++;
						}

						if( hasSiblingWithDifferentHue > 1 || hasHighContrastSibling > 1){
							return true;
						}
					}
				}
			}

			if(hasEquivilantSibling < 2){
				return true;
			}

			return false;
		}

		function errorPixel(px, offset, data1, data2){
			var data = errorPixelTransformer(data1, data2);
			px[offset] = data.r;
			px[offset + 1] = data.g;
			px[offset + 2] = data.b;
			px[offset + 3] = data.a;
		}

		function copyPixel(px, offset, data){
			px[offset] = data.r; //r
			px[offset + 1] = data.g; //g
			px[offset + 2] = data.b; //b
			px[offset + 3] = data.a * pixelTransparency; //a
		}

		function copyGrayScalePixel(px, offset, data){
			px[offset] = data.brightness; //r
			px[offset + 1] = data.brightness; //g
			px[offset + 2] = data.brightness; //b
			px[offset + 3] = data.a * pixelTransparency; //a
		}

		function getPixelInfo(data, offset, cacheSet){
			var r;
			var g;
			var b;
			var d;
			var a;

			r = data[offset];

			if(typeof r !== 'undefined'){
				g = data[offset+1];
				b = data[offset+2];
				a = data[offset+3];
				d = {
					r: r,
					g: g,
					b: b,
					a: a
				};

				return d;
			} else {
				return null;
			}
		}

		function addBrightnessInfo(data){
			data.brightness = getBrightness(data.r,data.g,data.b); // 'corrected' lightness
		}

		function addHueInfo(data){
			data.h = getHue(data.r,data.g,data.b);
		}

		function analyseImages(img1, img2, width, height){

			var data1 = img1.data;
			var data2 = img2.data;

      //TODO
      var imgd = new PNG({
          width: img1.width,
          height: img1.height,
          deflateChunkSize: img1.deflateChunkSize,
          deflateLevel: img1.deflateLevel,
          deflateStrategy: img1.deflateStrategy,
        });
			var targetPix = imgd.data;

			var mismatchCount = 0;

			var time = Date.now();

			var skip;

      var currentRectangle = null;
      var rectagnlesIdx = 0;

			if(!!largeImageThreshold && ignoreAntialiasing && (width > largeImageThreshold || height > largeImageThreshold)){
				skip = 6;
			}

			loop(height, width, function(verticalPos, horizontalPos){

				if(skip){ // only skip if the image isn't small
					if(verticalPos % skip === 0 || horizontalPos % skip === 0){
						return;
					}
				}

				var offset = (verticalPos*width + horizontalPos) * 4;
				var pixel1 = getPixelInfo(data1, offset, 1);
				var pixel2 = getPixelInfo(data2, offset, 2);

				if(pixel1 === null || pixel2 === null){
					return;
				}

        if (ignoreRectangles) {
          for(rectagnlesIdx = 0; rectagnlesIdx < ignoreRectangles.length; rectagnlesIdx++) {
            currentRectangle = ignoreRectangles[rectagnlesIdx];
            //console.log(currentRectangle, verticalPos, horizontalPos);
            if (
              (verticalPos >= currentRectangle[1]) &&
              (verticalPos < currentRectangle[1] + currentRectangle[3]) &&
              (horizontalPos >= currentRectangle[0]) &&
              (horizontalPos < currentRectangle[0] + currentRectangle[2])
            ) {
						  copyGrayScalePixel(targetPix, offset, pixel2);
					    //copyPixel(targetPix, offset, pixel1, pixel2);
              return;
            }
          }
        }

				if (ignoreColors){

					addBrightnessInfo(pixel1);
					addBrightnessInfo(pixel2);

					if( isPixelBrightnessSimilar(pixel1, pixel2) ){
						copyGrayScalePixel(targetPix, offset, pixel2);
					} else {
						errorPixel(targetPix, offset, pixel1, pixel2);
						mismatchCount++;
					}
					return;
				}

				if( isRGBSimilar(pixel1, pixel2) ){
					copyPixel(targetPix, offset, pixel1, pixel2);

				} else if( ignoreAntialiasing && (
						addBrightnessInfo(pixel1), // jit pixel info augmentation looks a little weird, sorry.
						addBrightnessInfo(pixel2),
						isAntialiased(pixel1, data1, 1, verticalPos, horizontalPos, width) ||
						isAntialiased(pixel2, data2, 2, verticalPos, horizontalPos, width)
					)){

					if( isPixelBrightnessSimilar(pixel1, pixel2) ){
						copyGrayScalePixel(targetPix, offset, pixel2);
					} else {
						errorPixel(targetPix, offset, pixel1, pixel2);
						mismatchCount++;
					}
				} else {
					errorPixel(targetPix, offset, pixel1, pixel2);
					mismatchCount++;
				}

			});

			data.misMatchPercentage = (mismatchCount / (height*width) * 100).toFixed(2);
			data.analysisTime = Date.now() - time;

			data.getDiffImage = function(text){
        return imgd;
			};
		}

		function compare(one, two){

			function onceWeHaveBoth(img){
				var width;
				var height;

        images.push(img);
				if(images.length === 2){
					width = images[0].width > images[1].width ? images[0].width : images[1].width;
					height = images[0].height > images[1].height ? images[0].height : images[1].height;

					if( (images[0].width === images[1].width) && (images[0].height === images[1].height) ){
						data.isSameDimensions = true;
					} else {
						data.isSameDimensions = false;
					}

					data.dimensionDifference = { width: images[0].width - images[1].width, height: images[0].height - images[1].height };

          //lksv: normalization removed
					analyseImages( images[0], images[1], width, height);

					triggerDataUpdate();
				}
			}

			images = [];
			loadImageData(one, onceWeHaveBoth);
			loadImageData(two, onceWeHaveBoth);
		}

		function getCompareApi(param){

			var secondFileData,
				hasMethod = typeof param === 'function';

			if( !hasMethod ){
				// assume it's file data
				secondFileData = param;
			}

			var self = {
				ignoreNothing: function(){

					tolerance.red = 16;
					tolerance.green = 16;
					tolerance.blue = 16;
					tolerance.alpha = 16;
					tolerance.minBrightness = 16;
					tolerance.maxBrightness = 240;

					ignoreAntialiasing = false;
					ignoreColors = false;

					if(hasMethod) { param(); }
					return self;
				},
				ignoreAntialiasing: function(){

					tolerance.red = 32;
					tolerance.green = 32;
					tolerance.blue = 32;
					tolerance.alpha = 32;
					tolerance.minBrightness = 64;
					tolerance.maxBrightness = 96;

					ignoreAntialiasing = true;
					ignoreColors = false;

					if(hasMethod) { param(); }
					return self;
				},
				ignoreColors: function(){

					tolerance.alpha = 16;
					tolerance.minBrightness = 16;
					tolerance.maxBrightness = 240;

					ignoreAntialiasing = false;
					ignoreColors = true;

					if(hasMethod) { param(); }
					return self;
				},
        //array of rectangles, each rectangle is defined as (x, y, width. height)
        //e.g. [[325, 170, 100, 40]]
        ignoreRectangles: function(rectangles) {
          ignoreRectangles = rectangles;
          return self;
        },
				repaint: function(){
					if(hasMethod) { param(); }
					return self;
				},
				onComplete: function( callback ){

					updateCallbackArray.push(callback);

					var wrapper = function(){
						compare(fileData, secondFileData);
					};

					wrapper();

					return getCompareApi(wrapper);
				}
			};

			return self;
		}

		return {
			onComplete: function( callback ){
				updateCallbackArray.push(callback);
				loadImageData(fileData, function(imageData, width, height){
					parseImage(imageData.data, width, height);
				});
			},
			compareTo: function(secondFileData){
				return getCompareApi(secondFileData);
			}
		};

	};

	_this['resemble'].outputSettings = function(options){
		var key;
		var undefined;

		if(options.errorColor){
			for (key in options.errorColor) {
				errorPixelColor[key] = options.errorColor[key] === undefined ? errorPixelColor[key] : options.errorColor[key];
			}
		}

		if(options.errorType && errorPixelTransform[options.errorType] ){
			errorPixelTransformer = errorPixelTransform[options.errorType];
		}
		
		pixelTransparency = options.transparency || pixelTransparency;

		if (options.largeImageThreshold !== undefined) {
			largeImageThreshold = options.largeImageThreshold;
		}

		return this;
	};

module.exports = _this['resemble']
